flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/table/UpsertKafkaTableITCase.java

```diff
@@ -0,0 +1,385 @@
+		expected.put(Row.of("boy"), Collections.singletonList(Row.ofKind(RowKind.INSERT, "boy", 1L)));
+
+		// ------------- create table ---------------
+
+		final String createSource = String.format(
+				"CREATE TABLE words_%s (" +
+						"  `word` STRING" +
+						") WITH (" +
+						"  'connector' = 'values'," +
+						"  'data-id' = '%s'" +
+						")",
+				format,
+				TestValuesTableFactory.registerData(inputData)
+		);
+		tEnv.executeSql(createSource);
+		final String createSinkTable = String.format(
+				"CREATE TABLE word_count_%s (\n"
+						+ "  `word` STRING,\n"
+						+ "  `count` BIGINT,\n"
+						+ "  PRIMARY KEY (`word`) NOT ENFORCED\n"
+						+ ") WITH (\n"
+						+ "  'connector' = 'upsert-kafka',\n"
+						+ "  'topic' = '%s',\n"
+						+ "  'properties.bootstrap.servers' = '%s',\n"
+						+ "  'key.format' = '%s',\n"
+						+ "  'value.format' = '%s'"
+						+ ")",
+				format,
+				topic,
+				bootstraps,
+				format,
+				format);
+		tEnv.executeSql(createSinkTable);
+		String initialValues = "INSERT INTO word_count_" + format + " " +
+				"SELECT LOWER(`word`), count(*) as `count` " +
+				"FROM words_" + format + " " +
+				"GROUP BY LOWER(`word`)";
+		tEnv.executeSql(initialValues).await();
+
+		// ---------- consume stream from sink -------------------
+
+		final List<Row> result = collectRows(tEnv.sqlQuery("SELECT * FROM word_count_" + format), 11);
+		comparedWithKeyAndOrder(expected, result, new int[]{0});
+	}
+
+	private void wordFreqToUpsertKafka() throws Exception {
+		// ------------- test data ---------------
+
+		final List<String> expectedData = Arrays.asList("3,1", "2,1");
+
+		// ------------- create table ---------------
+
+		final String createSinkTable = "CREATE TABLE sink_" + format + "(\n" +
+				"  `count` BIGINT,\n" +
+				"  `freq` BIGINT,\n" +
+				"  PRIMARY KEY (`count`) NOT ENFORCED\n" +
+				") WITH (\n" +
+				"  'connector' = 'values',\n" +
+				"  'sink-insert-only' = 'false'\n" +
+				")";
+		tEnv.executeSql(createSinkTable);
+		final TableResult query = tEnv.executeSql(
+				"INSERT INTO sink_" + format + "\n" +
+						"SELECT `count`, count(*) as `freq`\n" +
+						"FROM word_count_" + format + "\n" +
+						"GROUP BY `count`\n" +
+						"having count(*) < 2"
+		);
+
+		// ---------- consume stream from sink -------------------
+
+		waitingExpectedResults("sink_" + format, expectedData, Duration.ofSeconds(10));
+		query.getJobClient().get().cancel();
+	}
+
+	@Test
+	public void writeChangelogToUpsertKafka() throws Exception {
+		final String topic = USERS_TOPIC + "_" + format;
+		createTopic(topic);
+		String bootstraps = standardProps.getProperty("bootstrap.servers");
+
+		// ------------- test data ---------------
+
+		// Currently delete data may cause test unstable. We just delete the delete records.
+		List<Row> changelogData = Arrays.asList(
+				Row.ofKind(RowKind.INSERT, 100L, "Bob", "Beijing", Timestamp.valueOf("2020-08-15 00:00:01").toLocalDateTime()),
```

Comment:
I would prefer to use `TestValuesTableFactory#changelogRow` utility to create a changelog Row which is much shorter.  Besides, use `LocalDateTime.parse` instead of `Timestamp.valueOf`.

```java
changelogRow("+I", 100L, "Bob", "Beijing", LocalDateTime.parse("2020-08-15T00:00:01"))
```